add export-scala-regmap

albertchen-sifive opened the pull request:
adds a `duh-export-scala` script that can take a `duh` document and outputs scala files containing `RegisterRouter`s corresponding to the `memoryMaps` defined in the `duh` document.

example:
memMap:
```javascript
csrMemMap: {
  name: 'CSR',
  addressBlocks: [{
    name: 'csrAddressBlock',
    baseAddress: 0,
    range: 1024, width: 32,
    usage: 'register',
    volatile: false,
    registers: [{
      name: 'ODATA',
      addressOffset: 0, size: 64,
      displayName: 'Output Data Register',
      fields: [
        {name: 'data1', bits: 32, access: 'read-write', desc: 'this is the data1 field of ODATA'},
        {name: 'data2', bits: 32, access: 'write-only', desc: 'this is the data2 field of ODATA'},
      ]
    }, {
      name: 'OENABLE', addressOffset: 8, size: 32,
      displayName: 'Data direction',
      description: 'determines whether the pin is an input or an output. If the data direction bit is a 1, then the pin is an input',
      fields: [
        {name: 'foo', bits: 32, access: 'read-writeOnce', desc: 'this is the foo field of OENABLE'},
        {name: 'pad', bits: 32},
        {name: 'bar', bits: 32, access: 'writeOnce', desc: 'this is the bar field of OENABLE'},
      ]
    }, {
      name: 'IDATA', addressOffset: 12, size: 32,
      displayName: 'Input data',
      description: 'read the port pins',
      fields: [{name: 'data', bits: 32, access: 'read-only', desc: 'this is the data field of IDATA'}]
    }]
  }]
}
```

user scala:
```scala
package sifive.blocks.pio.CSR.csrAddressBlock

import chisel3._

object csrAddressBlockRegRouterUser {
  def deviceTreeName: String = csrAddressBlockRegRouterBase.deviceTreeName
  def deviceTreeCompat: Seq[String] = csrAddressBlockRegRouterBase.deviceTreeCompat

  object resetValues {
    object ODATA {
      def data: UInt =
        csrAddressBlockRegRouterBase.resetValues.ODATA.data
    }

    object OENABLE {
      def data: UInt =
        csrAddressBlockRegRouterBase.resetValues.OENABLE.data
    }

    object IDATA {
      def data: UInt =
        csrAddressBlockRegRouterBase.resetValues.IDATA.data
    }
  }
}
```

base scala:
```scala
// Generated Code
// Please DO NOT EDIT


package sifive.blocks.pio.CSR.csrAddressBlock

import chisel3._

import freechips.rocketchip.config.Parameters
import freechips.rocketchip.diplomacy.LazyModuleImp
import freechips.rocketchip.regmapper._
import freechips.rocketchip.tilelink.HasTLControlRegMap
import freechips.rocketchip.amba.axi4.HasAXI4ControlRegMap

class ODATARegisterBundle extends Bundle {
  val data = Output(UInt(32.W))
}

class OENABLERegisterBundle extends Bundle {
  val data = Output(UInt(32.W))
}

class IDATARegisterBundle extends Bundle {
  val data = Input(UInt(32.W))
}

class csrAddressBlockAddressBlockBundle extends Bundle {
  val ODATA = new ODATARegisterBundle()
  val OENABLE = new OENABLERegisterBundle()
  val IDATA = new IDATARegisterBundle()
}

object csrAddressBlockRegRouterBase {
  def deviceTreeName: String = "pio-csrAddressBlock"
  def deviceTreeCompat: Seq[String] = Seq("sifive,pio-0.1.0")

  object resetValues {
    object ODATA {
      def data: UInt = 0.U
    }

    object OENABLE {
      def data: UInt = 0.U
    }

    object IDATA {
      def data: UInt = 0.U
    }
  }
}

abstract class csrAddressBlockRegRouter(busWidthBytes: Int, baseAddress: Long)(implicit p: Parameters)
  extends IORegisterRouter(
    RegisterRouterParams(
      name = csrAddressBlockRegRouterUser.deviceTreeName,
      compat = csrAddressBlockRegRouterUser.deviceTreeCompat,
      base = baseAddress,
      beatBytes = busWidthBytes),
    new csrAddressBlockAddressBlockBundle) {

  lazy val module = new LazyModuleImp(this) {
    val portValue = ioNode.makeIO()
    val resetValue = Wire(portValue.cloneType.asOutput)
    resetValue.ODATA.data := csrAddressBlockRegRouterUser.resetValues.ODATA.data
    resetValue.OENABLE.data := csrAddressBlockRegRouterUser.resetValues.OENABLE.data
    resetValue.IDATA.data := csrAddressBlockRegRouterUser.resetValues.IDATA.data

    val register = RegInit(resetValue)
    portValue <> register

    val mapping = Seq(
      0 -> RegFieldGroup("ODATA", None, Seq(
        RegField(32, resetValue.ODATA.data, RegFieldDesc("data", "")))),
      4 -> RegFieldGroup("OENABLE", Some("determines whether the pin is an input or an output. If the data direction bit is a 1, then the pin is an input"), Seq(
        RegField(32, resetValue.OENABLE.data, RegFieldDesc("data", "")))),
      8 -> RegFieldGroup("IDATA", Some("read the port pins"), Seq(
        RegField.r(32, resetValue.IDATA.data, RegFieldDesc("data", "")))))
    regmap(mapping:_*)
  }
}

class csrAddressBlockTLRegMap(busWidthBytes: Int, baseAddress: Long)(implicit p: Parameters)
  extends csrAddressBlockRegRouter(busWidthBytes, baseAddress) with HasTLControlRegMap

class csrAddressBlockAXI4RegMapAXI4RegMap(busWidthBytes: Int, baseAddress: Long)(implicit p: Parameters)
  extends csrAddressBlockRegRouter(busWidthBytes, baseAddress) with HasAXI4ControlRegMap
```

albertchen-sifive committed:
add export-scala-regmap

albertchen-sifive committed:
fix typo

albertchen-sifive committed:
fix output path

albertchen-sifive committed:
fix lint errors

albertchen-sifive committed:
fix writeOnce, write-only

albertchen-sifive commented:
some TODOs:

- reset values? @drom these don't seem to be described by duh-schema? everything is reset to zero currently
- parameters? @drom duh-schema only allows integers for register widths
- `writeOnce`/`read-writeOnce`? @drom @terpstra not really sure what this means exactly. this is my best guess for the `RegWriteFn`
```
    val written = RegInit(false.B)
    RegWriteFn((valid, data) => {
      when (valid && !written) {
        register := data
        written := true.B
      }
      true.B
    })
```

albertchen-sifive committed:
add bundle directions, make and attach IO

drom reviewed (approved):
LGTM

richardxia commented on bin/duh-export-regmap.js:
```diff
@@ -0,0 +1,71 @@
+  const cwd = process.cwd();
+  const folderName = path.basename(cwd);
+  const fileName = argv._[0] || folderName + '.json5';
+  flow(Object.assign({filename: fileName}, argv));
```
I apologize in advance for any dumb questions I have about NodeJS, since I'm mostly familiar with frontend JS, not backend JS.

I'm trying to understand why `flow()` is being defined as a function that returns a Promise. This `main` function doesn't seem to do anything with the return value of `flow()`, meaning that the Promise is just being thrown away. I'm not sure I understand why `flow()` couldn't have been written as a normal function without wrapping a Promise.

drom replied:
It wrapped in Promise because it has async code in it. In case there are some actions after flow call you would need to wait for it.

richardxia replied:
I guess I am commenting on the fact that we don't actually wait for it in the current implementation, and because `flow` is not exported, no other files can import it and wait on it. To me, it feels like writing it as a Promise is extra overhead that doesn't provide value, and instead adds noise to understanding how this function works.

richardxia commented on bin/duh-export-regmap.js:
```diff
@@ -0,0 +1,71 @@
+    .then(duhCore.expandAll)
+    .then(duh1 => {
+      const emittedRegMappers = lib.exportScalaRegMap(duh1.component);
+      traverseSpec({
```
This is mostly a comment about `lib/traverse-spec.js` rather than its usage here. The lack of any comments for this function's definition makes it more work on the reader's end to understand exactly what it's doing, what its arguments are, and what the arguments of its (function-type) arguments are. Not blocking this PR, but can we add comments to that file to better describe the API of that function?

Also, from its name, I would guess that `traverseSpec()` is meant to be used on DUH document specs, but here you're calling it on an internal data structure that is not part of the user-facing DUH input JSON files. This is a code smell to me, and it suggests that either we are using `traverseSpec()` on an argument that is beyond the scope of what it should officially support or that the name of `traverseSpec()` is too specific and that it should be renamed to something more general, like `traverseObjectTree()`. Either case, having a documenting comment on the function describing the scope of its behavior would help with reader comprehension.

drom replied:
Agree, that `traverseSpec` need some comments. It has nothing to do with this PR.

richardxia commented on bin/duh-export-regmap.js:
```diff
@@ -0,0 +1,71 @@
+  if (argv.verbose) console.log('regmap');
+  duhCore.readDuh(argv)
+    .then(duhCore.expandAll)
+    .then(duh1 => {
```
Why is this argument named `duh1`? I would normally guess that the suffix `1` is for avoiding shadowing another variable of the same name, but I don't see another variable in this file named `duh`.

albertchen-sifive replied:
forgot to rename when I copied from another file with multiple `duh` variables

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+};
+
+
+module.exports = comp => {
```
I have a concern about directly assigning anonymous functions to `module.exports`. Doing it this way means that the anonymous function really has no name, which also makes it more difficult to use developer tooling such as text editor jumping to a definition (it at least defeats my ctags, which is not smart enough to understand this).

Could we adopt a convention of always assigning any exported function/class/variable to a name before exporting it? I think this would also help people who are less familiar with JavaScript read the codebase.

drom replied:
it is not a function but fat arrow function. It has no name as such.

richardxia replied:
Yes, I am familiar with the concept of fat arrows and understand that they don't technically have names. That said, there is a lot of existing JS tooling that understands that when you do a `const foo = () => {}`, that `foo` is the "name" of the function.

I'm going to make the argument that most programmers are used to searching a codebase for a function's name to find its definition, and by not first assigning it to a variable that roughly corresponds to the "name" of the function, people are not going to be able to find it as easily.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+  let result = {};
+  const packageNameBase = `package ${comp.vendor}.${comp.library}.${comp.name}`;
+  memoryMaps.forEach((memoryMap) => {
+    const addressBlocks = memoryMap.addressBlocks || [];
```
I don't know what our general approach is for handling invalid data, but according to `duh-schema`, the [`addressBlocks` property is required]( https://github.com/sifive/duh-schema/blob/0ca40f37d721d506256a22a6ca2899c6a543d8c3/lib/component.js#L130), making the `||` unnecessary.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+${indent(2)(regFieldSeqElements.join(',\n'))}))`;
+};
+
+const generateBody = addressBlock => packageName => {
```
Is there a reason you defined this as a curried function? At the only place where you actually call the function, you call it with both arguments simultaneously. I think it's much more typical in JavaScript to see multiple-parameter functions be defined as a single parameter list with multiple parameters rather than multiple, nested functions each taking a single parameter. Just to be absolutely clear, I am suggesting that we replace this with:

```suggestion
const generateBody = (addressBlock, packageName) => {
```

The same comment goes for most of your curried functions.

albertchen-sifive replied:
I just like curried functions better. couldn't hurt right?

drom replied:
we follow the same curry pattern for multiple functions with this scenario.
The reason is that we bind to the `addressBlock` first and then we can call in on multiple `packageName` in this address block.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+const generateBody = addressBlock => packageName => {
+  const scalaRegMapBundleName = 'register';
+
+  const registerMapElements = addressBlock.registers.map((register) => {
```
The addressBlock schema says that the [`registers` property is optional](https://github.com/sifive/duh-schema/blob/0ca40f37d721d506256a22a6ca2899c6a543d8c3/lib/component.js#L137), so don't you need to check if this is defined before mapping over it?

drom replied:
Good catch.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+const generateRegister = register => scalaRegMapBundleName => {
+  const scalaRegBundleName = `${scalaRegMapBundleName}.${register.name}`;
+
+  const regFieldSeqElements = register.fields.map((field, idx) => {
```
The schema says that the [`fields` property is optional](https://github.com/sifive/duh-schema/blob/0ca40f37d721d506256a22a6ca2899c6a543d8c3/lib/register.js#L7), so don't you need to check that it exists first?

drom replied:
Yes, we must check it first.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+//// GNERATE REGROUTER ////
+
+const generateRegisterField = field => scalaRegisterName => {
+  if (field.access) {
```
As far as I can tell from the schema, the [`access` property is defined on the register](https://github.com/sifive/duh-schema/blob/0ca40f37d721d506256a22a6ca2899c6a543d8c3/lib/register.js#L12), not the field.

richardxia commented on lib/export-scala-regmap.js:
```diff
@@ -0,0 +1,184 @@
+  if (field.access) {
+    const regFieldDesc = field.name || field.desc ?
+      `, RegFieldDesc(${field.name ? `"${field.name}"` : '""'}, ${field.desc ? `"${field.desc}"` : '""'}))` :
+      ')';
```
I'm not a big fan of terminating a parenthesis that is defined outside of this string, since it's unintuitive to have to look elsewhere to understand what it's closing. Can we just put the terminating parenthesis for each of the `RegField(` calls in the same string that opens the parenthesis?

I'd also like to make a similar comment about the leading comma, even though I do understand that it's for making the final argument optional and not having extra trailing commas in the generated Scala code. It still makes it more difficult to read and it's not immediately obvious what's going on. On my first read I thought that `scalaRegisterName` was getting string concatenated with `regFieldDesc` as a way to prefix/namespace something, but I had to read more carefully to understand that we were trying to avoid a trailing comma.

albertchen-sifive committed:
assign regmap export to variable

albertchen-sifive committed:
don't use curried funtions

albertchen-sifive committed:
rename duh1 to duh

albertchen-sifive committed:
don't provide default for required addresssBlocks field

albertchen-sifive committed:
get access from register instead of register field

albertchen-sifive committed:
make regFieldDesc string interpolation easier to read

albertchen-sifive committed:
provide default if register fields is undefined

albertchen-sifive committed:
fix undefined variables

albertchen-sifive committed:
default to addressBlock access if undefined in register

albertchen-sifive committed:
fix switch indentation

albertchen-sifive committed:
do not enforce writeOnce

albertchen-sifive committed:
add DO NOT EDIT to top of generated regmap scala

albertchen-sifive committed:
move defaults into object

albertchen-sifive committed:
organize names

albertchen-sifive committed:
split out parameters into user and base objects

albertchen-sifive committed:
split out reset assignments

albertchen-sifive committed:
fix lint issues

albertchen-sifive committed:
add overrideable reset values

albertchen-sifive committed:
don't overwrite user files

albertchen-sifive committed:
more name organization

albertchen-sifive committed:
update for new reg schema with reset values

albertchen-sifive commented:
updated after feedback from @terpstra and @drom 
- pulled parameters into a `object` with overrides that live in an `object` in the user code
- better default values from device tree compat and name strings
- ignore `writeOnce`/`read-writeOnce` and just emit a `write-only`/`read-write` register since write-once can be done in software
- added user-overridable reset values for registers with defaults that are pulled from the duh document
- works with the new `0.6.0` schema that allows specifying `access` and `resetValue` for register fields.

richardxia reviewed (approved):
I don't think I can give a thorough review, but what I can see looks good to me.
